examples/throughput/publisher.cpp

```diff
@@ -0,0 +1,223 @@
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+#include <string>
+#include <vector>
+#include <thread>
+#include <chrono>
+#include <csignal>
+
+#include "dds/dds.hpp"
+#include "Throughput.hpp"
+
+/*
+ * The Throughput example measures data throughput in bytes per second. The publisher
+ * allows you to specify a payload size in bytes as well as allowing you to specify
+ * whether to send data in bursts. The publisher will continue to send data forever
+ * unless a time out is specified. The subscriber will receive data and output the
+ * total amount received and the data rate in bytes per second. It will also indicate
+ * if any samples were received out of order. A maximum number of cycles can be
+ * specified and once this has been reached the subscriber will terminate and output
+ * totals and averages.
+ */
+
+using namespace org::eclipse::cyclonedds;
+
+#define MAX_SAMPLES 100
+#define pubprefix "=== [Publisher] "
+
+static volatile sig_atomic_t done(false); /*semaphore for keeping track of whether to run the test*/
+
+static uint32_t payloadSize(8192); /*size of the payload of the sent messages*/
+
+static std::chrono::milliseconds burstInterval(100); /*interval between bursts of messages*/
+
+static uint32_t burstSize(10); /*number of messages to send each burst*/
+
+static std::chrono::seconds timeOut(30); /*timeout before the writer will give up*/
+
+static std::string partitionName("Throughput example"); /*name of the domain on which the throughput test is run*/
+
+static int parse_args(
+    int argc,
+    char **argv)
+{
+  int result = EXIT_SUCCESS;
+  /*
+   * Get the program parameters
+   * Parameters: publisher [payloadSize] [burstInterval] [burstSize] [timeOut] [partitionName]
+   */
+  if (argc == 2 && (strcmp (argv[1], "-h") == 0 || strcmp (argv[1], "--help") == 0))
+  {
+    std::cout << pubprefix << "Usage (parameters must be supplied in order):\n";
+    std::cout << pubprefix << "./publisher [payloadSize (bytes)] [burstInterval (ms)] [burstSize (samples)] [timeOut (seconds)] [partitionName]\n";
+    std::cout << pubprefix << "Defaults:\n";
+    std::cout << pubprefix << "./publisher 8192 0 1 0 \"Throughput example\"\n" << std::flush;
+    return EXIT_FAILURE;
+  }
+
+  if (argc > 1)
+    payloadSize = static_cast<uint32_t>(atoi (argv[1])); /* The size of the payload in bytes */
+
+  if (argc > 2)
+    burstInterval = std::chrono::milliseconds(atoi (argv[2])); /* The time interval between each burst in ms */
+
+  if (argc > 3)
+    burstSize = static_cast<uint32_t>(atoi (argv[3])); /* The number of samples to send each burst */
+
+  if (argc > 4)
+    timeOut = std::chrono::seconds(atoi (argv[4])); /* The number of seconds the publisher should run for (0 = infinite) */
+
+  if (argc > 5)
+    partitionName = argv[5]; /* The name of the partition */
```

Comment:
I used getopt for the omg interopt testing. But that is hidden away, and there are many more arguments to parse (of different types). I think for such a small example (a few numbers), this style is okay. I don't see much added value to having a full fledge CLI here.

Reply:
I did not use an option list compared to fixed parameter ordering because my approach was to have the C++ implementation be completely exchangeable with the C implementation. So that means no option list as command line parameters.